### PR TITLE
Add edges

### PR DESCRIPTION
**Pull request by WhiteHalmos**
Split change into 3+ commits to make it easier to review.
* First commit is #857, will be rebased out once that is merged.
* Second commit is #888, has no logic change, paves the way for edges.
* Third commit and onwards adds edges (16 bytes) to replace the vast majority of nodes (56 bytes).

Playouts | Nodes (KB) | Nodes (MB) | n/s | Edges (KB) | Edges (MB) | n/s | Mem | Speed
-- | -- | -- | -- | -- | -- | -- | -- | --
1000 | 158928 | 155 | 1063 | 150324 | 147 | 1086 | -5.41% | 2.16%
5000 | 263584 | 257 | 1213 | 209280 | 204 | 1262 | -20.60% | 4.04%
10000 | 391964 | 383 | 1312 | 284100 | 277 | 1360 | -27.52% | 3.66%
50000 | 1419948 | 1387 | 1350 | 886796 | 866 | 1392 | -37.55% | 3.11%
100000 | 2692292 | 2629 | 1326 | 1625284 | 1587 | 1394 | -39.63% | 5.13%

<details>
<summary>Test details (default 1600)</summary>

next (with benchmark PR) sizeof(UCTNode) == 56
```
/usr/bin/time -v ./leelaz-next -w w --benchmark            
Thinking at most 2678400.0 seconds...                                                                            
NN eval=0.476396                                        

  D4 ->     511 (V: 50.36%) (N: 28.82%) PV: D4 D16 Q4 C3 C4 D3 E3 E2 F2 F3 E4 G2 D2 F1 C2 R3 R4 Q3 P3 P2 O2 O3 P4
 N2 Q2 O1                                               
  Q4 ->     418 (V: 50.25%) (N: 25.25%) PV: Q4 D4 D16 R3 Q3 R4 R5 S5 S6 R6 Q5 S7 S4 T6 S3 C17 D17 C16 C15 B15 B14
 C14 D15                                                
 D16 ->     389 (V: 50.23%) (N: 23.48%) PV: D16 Q4 D4 C17 D17 C16 C15 B15 B14 C14 D15 B13 B16 A14 B17 C3 D3 C4 C5
 B5 B6 C6 D5                                            
  R4 ->      57 (V: 50.09%) (N:  3.76%) PV: R4 D4 D16 C17 D17 C16 C15
  Q3 ->      51 (V: 50.32%) (N:  2.86%) PV: Q3 D4 D16 C17 C16 D17 E17 E18
 C16 ->      44 (V: 50.43%) (N:  2.41%) PV: C16 D4 Q4 R3 R4 Q3 P3 P2
 D17 ->      36 (V: 50.11%) (N:  2.28%) PV: D17 Q4 D4 C3 D3 C4 C5
  D3 ->      34 (V: 50.04%) (N:  2.21%) PV: D3 D16 Q4 R3 Q3 R4
 R17 ->      32 (V: 49.13%) (N:  3.06%) PV: R17 Q17 R16 R15 S15 S14 R14 Q15 S13 S16 T14 S17 Q4 D16 D4 R3
  C4 ->      28 (V: 49.91%) (N:  2.03%) PV: C4 Q4 D16 C17 D17 C16
1601 visits, 563325 nodes, 1600 playouts, 1118 n/s

       Command being timed: "./leelaz-next -w w --benchmark"                                                    
        User time (seconds): 1.58                       
        System time (seconds): 0.57                      
        Percent of CPU this job got: 112%               
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:01.92                                                     
        Average shared text size (kbytes): 0            
        Average unshared data size (kbytes): 0          
        Average stack size (kbytes): 0                  
        Average total size (kbytes): 0                   
        Maximum resident set size (kbytes): 174788       
        Average resident set size (kbytes): 0            
        Major (requiring I/O) page faults: 0            
        Minor (reclaiming a frame) page faults: 27090   
        Voluntary context switches: 87                  
        Involuntary context switches: 46                
        Swaps: 0             
        File system inputs: 0                           
        File system outputs: 0                          
        Socket messages sent: 0                         
        Socket messages received: 0                     
        Signals delivered: 0
        Page size (bytes): 4096                         
        Exit status: 0
```

After sizeof(UCTEdge) == 16
```
/usr/bin/time -v ./leelaz-edges -w w --benchmark           
Thinking at most 2678400.0 seconds...                                                                            
NN eval=0.476396                                                                                                 

  D4 ->     511 (V: 50.36%) (N: 28.82%) PV: D4 D16 Q4 C3 C4 D3 E3 E2 F2 F3 E4 G2 D2 F1 C2 R3 R4 Q3 P3 P2 O2 O3 P4
 N2 Q2 O1                                               
  Q4 ->     418 (V: 50.25%) (N: 25.25%) PV: Q4 D4 D16 R3 Q3 R4 R5 S5 S6 R6 Q5 S7 S4 T6 S3 C17 D17 C16 C15 B15 B14
 C14 D15                                                
 D16 ->     389 (V: 50.23%) (N: 23.48%) PV: D16 Q4 D4 C17 D17 C16 C15 B15 B14 C14 D15 B13 B16 A14 B17 C3 D3 C4 C5
 B5 B6 C6 D5                                            
  R4 ->      57 (V: 50.09%) (N:  3.76%) PV: R4 D4 D16 C17 D17 C16 C15
  Q3 ->      51 (V: 50.32%) (N:  2.86%) PV: Q3 D4 D16 C17 C16 D17 E17 E18
 C16 ->      44 (V: 50.43%) (N:  2.41%) PV: C16 D4 Q4 R3 R4 Q3 P3 P2
 D17 ->      36 (V: 50.11%) (N:  2.28%) PV: D17 Q4 D4 C3 D3 C4 C5
  D3 ->      34 (V: 50.04%) (N:  2.21%) PV: D3 D16 Q4 R3 Q3 R4
 R17 ->      32 (V: 49.13%) (N:  3.06%) PV: R17 Q17 R16 R15 S15 S14 R14 Q15 S13 S16 T14 S17 Q4 D16 D4 R3
  C4 ->      28 (V: 49.91%) (N:  2.03%) PV: C4 Q4 D16 C17 D17 C16
1601 visits, 1601 nodes, 563325 edges, 1600 playouts, 1176 n/s

       Command being timed: "./leelaz-edges -w w --benchmark"                                                   
        User time (seconds): 1.46                       
        System time (seconds): 0.61                      
        Percent of CPU this job got: 112%               
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:01.84                                                     
        Average shared text size (kbytes): 0            
        Average unshared data size (kbytes): 0          
        Average stack size (kbytes): 0                  
        Average total size (kbytes): 0                   
        Maximum resident set size (kbytes): 157452       
        Average resident set size (kbytes): 0            
        Major (requiring I/O) page faults: 0            
        Minor (reclaiming a frame) page faults: 22724   
        Voluntary context switches: 86                  
        Involuntary context switches: 51                
        Swaps: 0             
        File system inputs: 0                           
        File system outputs: 0                          
        Socket messages sent: 0                         
        Socket messages received: 0                     
        Signals delivered: 0
        Page size (bytes): 4096                         
        Exit status: 0
```
</details>

### CODE REVIEW

**Comment by WhiteHalmos** [comment]
Benchmark, so after b plays q16.
```
/usr/bin/time -v ./leelaz-edges -w w --benchmark --noponder -p500000                                                                                                        
Thinking at most 2678400.0 seconds...                                                                            
NN eval=0.476396
...
  C4 ->  234346 (V: 50.79%) (N:  2.03%) PV: C4 D16 Q4 E3 D6 F17 R17 Q17 R16 R15 S15 S14 R14 Q15 S13 S16 T14 S17 G
3 C3 B3 B2 D3 C2 D4 D2 E4 F3 F4 G2 H2 H3 G4 J2 F2 H1    
  Q4 ->   93251 (V: 50.64%) (N: 25.25%) PV: Q4 D4 D16 R3 Q3 R4 R5 S5 S6 R6 Q5 S7 S4 T6 S3 C17 D17 C16 C15 B15 B14
 C14 D15 B13 B16 A14 B17 C18 B18 D14 C6 F4 B4 E15 F16 E16
  D4 ->   74609 (V: 50.57%) (N: 28.82%) PV: D4 C3 D3 C4 C5 B5 B6 C6 D5 B7 B4 A6 B3 C16 Q4 E16 R17 R16 Q17 P17 P18
 O18 O17 P16 N18 Q18 O19 R18 F16 F17 G17 G16 F15 H17 E17 G18 F18 G17 D16 C17 C15 E18 E15 D17 C11 R3
 D16 ->   62765 (V: 50.58%) (N: 23.48%) PV: D16 D4 Q4 R3 Q3 R4 R5 S5 S6 R6 Q5 S7 S4 T6 S3 C17 D17 C16 C15 B15 B14
 C14 D15 B13 B16 A14 B17 C18 B18 D14 C6 F4 B4 E15 F16                                                            
  R4 ->    8493 (V: 50.54%) (N:  3.76%) PV: R4 D4 D16 P3 R17 R16 Q17 P17 P18 O18 O17 P16 N18 Q18 O19 R18 C3 D3 C4
 C5 B5 B6 C6 D5 B7 B4 A6 B3 C2 B2 S17 S18 O16 Q5
  Q3 ->    7990 (V: 50.59%) (N:  2.86%) PV: Q3 D4 D16 R5 C3 C4 D3 E3 E2 F2 F3 E4 G2 D2 F1 C2 R17 R16 Q17 P17 P18
O18 O17 P16 N18 Q18 O19 R18 Q5 Q6 R4 P5 Q4 P7
 C16 ->    6169 (V: 50.57%) (N:  2.41%) PV: C16 D4 Q4 E17 C3 C4 D3 E3 E2 F2 F3 E4 G2 D2 F1 C2 R17 R16 Q17 P17 P18
 O18 O17 P16 N18 Q18 O19 R18 B3 B2 F4 D15 C15
  D3 ->    6157 (V: 50.59%) (N:  2.21%) PV: D3 Q4 D16 C5 D5 D6 C4 E5 D4 C17 D17 C16 C15 B15 B14 C14 D15 B13 B16 A
14 B17 C18 B18 D14 E6 D7 E7 D8
 D17 ->    5078 (V: 50.54%) (N:  2.28%) PV: D17 D4 Q4 C15 R17 R16 Q17 P17 P18 O18 O17 P16 N18 Q18 O19 R18 C3 C4 D
3 E3 E2 F2 F3 E4 G2 D2 F1 C2 B3 B2 C16 D15
 R17 ->    1118 (V: 49.24%) (N:  3.06%) PV: R17 Q17 R16 R15 S15 S14 R14 Q15 S13 S16 T14 S17 Q4 D16 D4 R3 R4 Q3 P3
 P2 O2 O3 P4 N2 Q2 O1 R2 C3 C4
 R16 ->      12 (V: 48.20%) (N:  0.06%) PV: R16 R17 R15 Q15 R14 Q17 R11 Q4
 Q17 ->      12 (V: 47.95%) (N:  0.06%) PV: Q17 R17 P17 P16 O17 R16 Q4 D16
500001 visits, 500001 nodes, 171074878 edges, 500000 playouts, 1017 n/s

        Command being timed: "./leelaz-edges -w w --benchmark --noponder -p500000"                               
        User time (seconds): 400.41                                                                              
        System time (seconds): 99.50                                                                             
        Percent of CPU this job got: 100%                                                                        
        Elapsed (wall clock) time (h:mm:ss or m:ss): 8:19.58                                                     
        Average shared text size (kbytes): 0                                                                     
        Average unshared data size (kbytes): 0          
        Average stack size (kbytes): 0                                                                           
        Average total size (kbytes): 0                   
        Maximum resident set size (kbytes): 7013788                                                              
        Average resident set size (kbytes): 0                                                      
        Major (requiring I/O) page faults: 10                                                                    
        Minor (reclaiming a frame) page faults: 1736948                                                          
        Voluntary context switches: 5868                                                                         
        Involuntary context switches: 710       
        Swaps: 0                                                                                                
        File system inputs: 2136             
        File system outputs: 0                                                                                   
        Socket messages sent: 0              
        Socket messages received: 0                                                                              
        Signals delivered: 0  
        Page size (bytes): 4096                                                                                  
        Exit status: 0
```

**Comment by sethtroisi** [comment]
I've been working making on @ywrt expand node PR which would save a small amount more memory but I don't think is as clean as this pull request. 

I'll do a full review later today but on a initial look this appears to avoid all the pitfalls I've experienced.

File: src/UCTSearch.h [hunk 1]
```diff
@@ -73,12 +73,12 @@ class UCTSearch {
     static constexpr passflag_t NORESIGN = 1 << 1;
 
     /*
-        Maximum size of the tree in memory. Nodes are about
-        48 bytes, so limit to ~1.2G on 32-bits and about 5.5G
-        on 64-bits.
+        Maximum size of the tree in memory. Nodes are 56 bytes, edges are
```
Comment (Hersmunch) [inline src/UCTSearch.h]: sizeof(UCTNode) == 64 for me. I count 60 bytes by adding up the variables so I believe there is either 4 bytes padding between m_has_edges and m_edges or there is spare at the end

**Comment by sethtroisi** [comment]
You can verify this is functional (output? deterministic? algorithmic?) no-op with
`echo -e "time_settings 0 1 0\nplay black q16\ngenmove white" | ./leelaz  -w weights --noponder -m0 -t1 -s1 --timemanage off -p5000`

determinism is good, functional no-op is good.

**Comment by zediir** [comment]
5000000 result is because you hit the node limit earlier with leelaz-next?

**Comment by sethtroisi** [comment]
@WhiteHalmos, I took over @ywrt Pull Request https://github.com/gcp/leela-zero/pull/574 and wrote https://github.com/gcp/leela-zero/pull/887, I'd be interested in what you think. It uses 6 bytes for each unexpanded child but comes with some extra code complexity

**Comment by WhiteHalmos** [comment]
Closing this one as I think @sethtroisi's PR #887 is much better. I opened another PR for the one useful commit in this one.

**Pull request closed without merging**